{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0d93a936",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "library(tidyverse)\n",
    "library(Gviz)\n",
    "library(rtracklayer)\n",
    "library(GenomicFeatures)\n",
    "library(data.table)\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08c494c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "GTF_dir <- \"../reference_files/gencode.v40.annotation.gtf\"\n",
    "\n",
    "\n",
    "## comment back in if you need to make txdb\n",
    "#hg38_txdb <- makeTxDbFromGFF(GTF_dir,\n",
    "#                                 format=\"gtf\",\n",
    "#                                 organism=\"Homo sapiens\",\n",
    "#                                 dbxrefTag = \"gene_name\")\n",
    " \n",
    " \n",
    "#saveDb(hg38_txdb, \"~/BE_paper_2022/reference_files/hg38.txdb\")\n",
    "\n",
    "hg38_txdb <- loadDb(\"../reference_files/hg38.txdb\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "85cc1774",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtrack <- GenomeAxisTrack()\n",
    "# specify the start and end of chromosome \n",
    "chr_no <- 12 \n",
    "chr_start <- 9760000  # start of region\n",
    "chr_end <- 9768000 # end of region\n",
    "\n",
    "gen <- 'hg38'\n",
    "\n",
    "## set gene region track\n",
    "gtTrack <- GeneRegionTrack(hg38_txdb,\n",
    "                           chromosome=chr_no, # chromosome number\n",
    "                           start=chr_start, # start of region\n",
    "                           end=chr_end, # end of region\n",
    "                           transcriptAnnotation=\"gene_id\", # symbol is the gene symbol\n",
    "                           fontsize.group=20# free to adjust font size\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "92f75780",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "bedgraph_enformer <- fread(\n",
    "  \"../data_and_analyses/Enformer_scores/CD69_original_enformer_Fig1D_locus/CD69.pooled.bedGraph.pooled.bedGraph\",\n",
    "  col.names = c('chromosome', 'start', 'end', 'value')\n",
    ")\n",
    "\n",
    "\n",
    "## here changed DA regions to bed format\n",
    "## here change da windows format to bed, then filter to keep only specific ROI\n",
    "system(\"tail -n+2 ../data_and_analyses/ATAC/GSE155555/stim_v_ctrl_csaw_DA-windows_all.txt | awk '{OFS=\\\"\\t\\\"}{print $1,$2,$3,$13}' > ../data_and_analyses/ATAC/GSE155555/stim_v_ctrl_csaw_DA-windows_all.bed\")\n",
    "system(\"awk '$4 <= 0.20' ../data_and_analyses/ATAC/GSE155555/stim_v_ctrl_csaw_DA-windows_all.bed > ../data_and_analyses/ATAC/GSE155555/stim_v_ctrl_csaw_DA-windows.filtq2.bed\")\n",
    "\n",
    "jurkat_DA <- import.bed(\"../data_and_analyses/ATAC/GSE155555/stim_v_ctrl_csaw_DA-windows.filtq2.bed\")\n",
    "guides_CD69 <- import.bed(\"../data_and_analyses/guides_and_perturbation_results/dCas9_KRAB_guides.bed\")\n",
    "\n",
    "ctrl_bw_jurkat = import.bw(\"../data_and_analyses/ATAC/GSE155555/bigwigs/Jurkat_WT_1_none_unspecified.merged.bw\",as=\"GRanges\")\n",
    "stim_bw_jurkat = import.bw(\"../data_and_analyses/ATAC/GSE155555/bigwigs/Jurkat_WT_1_anti_CD3_PMA_4h_unspecified.merged.bw\",as=\"GRanges\")\n",
    "ctrl_bw_cd4 = import.bw(\"../data_and_analyses/ATAC/GSE124867/CD4_resting.merged.bw\", as=\"GRanges\")\n",
    "stim_bw_cd4 = import.bw(\"../data_and_analyses/ATAC/GSE124867/CD4_activated.merged.bw\", as=\"GRanges\")\n",
    "\n",
    "options(ucscChromosomeNames=FALSE)\n",
    "IG_bedgraph <- DataTrack(\n",
    "  range = bedgraph_enformer,\n",
    "  type = \"a\",\n",
    "  genome = 'hg38',\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  name = \"Input*Gradient\",\n",
    "    fill.histogram = \"red\", \n",
    "    alpha=1\n",
    ")\n",
    "\n",
    "jurkat_DA_regions <- AnnotationTrack(\n",
    "  range = jurkat_DA,\n",
    "  chromosome=chr_no,\n",
    "  type = \"a\",\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end\n",
    ")\n",
    "\n",
    "\n",
    "CD69_guides <- AnnotationTrack(\n",
    "  range = guides_CD69,\n",
    "  chromosome=chr_no,\n",
    "  type = \"a\",\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end # end of region\n",
    ")\n",
    "\n",
    "\n",
    "jurkat_ctrl_track <- DataTrack(\n",
    "  range = ctrl_bw_jurkat,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\"\n",
    ")\n",
    "\n",
    "jurkat_stim_track <- DataTrack(\n",
    "  range = stim_bw_jurkat,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\"\n",
    ")\n",
    "\n",
    "cd4_ctrl_track <- DataTrack(\n",
    "  range = ctrl_bw_cd4,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\"\n",
    ")\n",
    "\n",
    "cd4_stim_track <- DataTrack(\n",
    "  range = stim_bw_cd4,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a3e71f86",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pdf(\"cd69_locus_plot_bigwigs.pdf\")\n",
    "## modify remainder in AI \n",
    "plotTracks(list(gtrack,\n",
    "                gtTrack,\n",
    "                cd4_ctrl_track,\n",
    "                cd4_stim_track,\n",
    "                jurkat_ctrl_track,\n",
    "                jurkat_stim_track,\n",
    "                jurkat_DA_regions,\n",
    "                CD69_guides,\n",
    "                IG_bedgraph), shape = \"arrow\", \n",
    "           transcriptAnnotation = \"symbol\", chromosome = \"chr12\",from = 9732000, to= 9776000,\n",
    "           type = c(\"h\",\"coverage\"),scale = 0.5, \n",
    "           labelPos = \"below\"\n",
    "           )\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52138cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "zoom_bedgraph_enformer <- fread(\n",
    "  \"~/BE_paper_2022/data/enformer_results/CD69.all.bedGraph.all.bedGraph\",\n",
    "  col.names = c('chromosome', 'start', 'end', 'value')\n",
    ")\n",
    "zoom_bedgraph_enformer$value <- abs(zoom_bedgraph_enformer$value)\n",
    "\n",
    "IG_bedgraph <- DataTrack(\n",
    "  range = zoom_bedgraph_enformer,\n",
    "  type = c(\"gradient\"),\n",
    "  genome = 'hg38',\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  name = \"Input*Gradient\",\n",
    "    col=\"red\",\n",
    "    showColorBar=TRUE\n",
    "    \n",
    ")\n",
    "pdf(\"cd69_gradient_9.60_9.68.pdf\")\n",
    "plotTracks(list(gtrack,\n",
    "                #gtTrack,\n",
    "                IG_bedgraph), shape = \"arrow\", \n",
    "           transcriptAnnotation = \"symbol\", chromosome = \"chr12\",from = 9760000, to= 9768000,\n",
    "           type = c(\"gradient\"),\n",
    "           ncolor=50,\n",
    "           gradient=c(\"white\",\"red\")\n",
    "           )\n",
    "dev.off()"
   ]
  }
 ],
 "metadata": {
  "environment": {
   "name": "r-cpu.4-0.m76",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-0:m76"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
