{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "022dc6a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# csaw workflow for ATAC-seq differential accessibility analysis,\n",
    "# adapted from Jake Reske, MSU 2020, reskejak@msu.edu, https://github.com/reskejak\n",
    "\n",
    "suppressPackageStartupMessages({library(GenomicRanges)\n",
    "library(csaw)\n",
    "library(\"edgeR\")\n",
    "library(data.table)\n",
    "library(ggplot2)\n",
    "library(dplyr)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "52d922e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in replicate narrowPeaks\n",
    "control1.peaks <- fread(\"peaks/P258-1_S1_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.tn5.pval0.01.300K.bfilt.narrowPeak.gz\", \n",
    "                           sep=\"\\t\",header=FALSE)[,1:3]\n",
    "control2.peaks <- fread(\"peaks/P258-2_S2_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.tn5.pval0.01.300K.bfilt.narrowPeak.gz\", \n",
    "                           sep=\"\\t\",header=FALSE)[,1:3]\n",
    "control3.peaks <- fread(\"peaks/P258-3_S3_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.tn5.pval0.01.300K.bfilt.narrowPeak.gz\", \n",
    "                           sep=\"\\t\",header=FALSE)[,1:3]\n",
    "treat1.peaks <- fread(\"peaks/P260-1_S9_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.tn5.pval0.01.300K.bfilt.narrowPeak.gz\", \n",
    "                           sep=\"\\t\",header=FALSE)[,1:3]\n",
    "treat2.peaks <- fread(\"peaks/P260-2_S10_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.tn5.pval0.01.300K.bfilt.narrowPeak.gz\", \n",
    "                           sep=\"\\t\",header=FALSE)[,1:3]\n",
    "treat3.peaks <- fread(\"peaks/P260-3_S11_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.tn5.pval0.01.300K.bfilt.narrowPeak.gz\", \n",
    "                           sep=\"\\t\",header=FALSE)[,1:3]\n",
    "colnames(control1.peaks) <- c(\"chrom\", \"start\", \"end\")\n",
    "colnames(control2.peaks) <- c(\"chrom\", \"start\", \"end\")\n",
    "colnames(control3.peaks) <- c(\"chrom\", \"start\", \"end\")\n",
    "colnames(treat1.peaks) <- c(\"chrom\", \"start\", \"end\")\n",
    "colnames(treat2.peaks) <- c(\"chrom\", \"start\", \"end\")\n",
    "colnames(treat3.peaks) <- c(\"chrom\", \"start\", \"end\")\n",
    "\n",
    "control1.peaks <- GRanges(control1.peaks)\n",
    "control2.peaks <- GRanges(control2.peaks)\n",
    "control3.peaks <- GRanges(control3.peaks)\n",
    "treat1.peaks <- GRanges(treat1.peaks)\n",
    "treat2.peaks <- GRanges(treat2.peaks)\n",
    "treat3.peaks <- GRanges(treat3.peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1e87e2cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# interesect b/w bio replicates, union across conditions\n",
    "treat.peaks <- GenomicRanges::intersect(treat1.peaks, treat2.peaks)\n",
    "treat.peaks <- GenomicRanges::intersect(treat.peaks,treat3.peaks)\n",
    "control.peaks <- GenomicRanges::intersect(control1.peaks, control2.peaks)\n",
    "control.peaks <- GenomicRanges::intersect(control.peaks,control3.peaks)\n",
    "all.peaks <- GenomicRanges::union(treat.peaks, control.peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "58f229c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify paired-end BAMs\n",
    "pe.bams <- c(\"bams/P258-1_S1_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\",\n",
    "             \"bams/P258-2_S2_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\",\n",
    "             \"bams/P258-3_S3_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\",\n",
    "             \"bams/P260-1_S9_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\",\n",
    "             \"bams/P260-2_S10_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\",\n",
    "             \"bams/P260-3_S11_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "\n",
    "blacklist <- read.table(\"../../../reference_files/hg38_blacklist_ENCODE_ENCFF356LFX.bed\", sep=\"\\t\")\n",
    "colnames(blacklist) <- c(\"chrom\", \"start\", \"end\")\n",
    "blacklist <- GRanges(blacklist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3040dbe3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define read parameters\n",
    "standard.chr <- paste0(\"chr\", c(1:22)) # only use standard chromosomes\n",
    "param <- readParam(max.frag=1000, pe=\"both\", discard=blacklist, restrict=standard.chr)\n",
    "\n",
    "##############################\n",
    "# count reads in windows specified by MACS2                                      \n",
    "peak.counts <- regionCounts(pe.bams, all.peaks, param=param)\n",
    "\n",
    "##############################\n",
    "# MACS2 peaks only: filter low abundance peaks\n",
    "library(\"edgeR\")\n",
    "peak.abundances <- aveLogCPM(asDGEList(peak.counts)) \n",
    "peak.counts.filt <- peak.counts[peak.abundances > -3, ] # only use peaks logCPM > -3\n",
    "# few or no peaks should be removed; modify as desired\n",
    "\n",
    "##############################\n",
    "\n",
    "# get paired-end fragment size distribution\n",
    "#control1.pe.sizes <- getPESizes(\"~/BE_paper_2022/data/ATAC/258_v_260_edit_v_WT/P258-1_S1_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "#control2.pe.sizes <- getPESizes(\"~/BE_paper_2022/data/ATAC/258_v_260_edit_v_WT/P258-2_S2_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "#control3.pe.sizes <- getPESizes(\"~/BE_paper_2022/data/ATAC/258_v_260_edit_v_WT/P258-3_S3_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "\n",
    "#treat1.pe.sizes <- getPESizes(\"~/BE_paper_2022/data/ATAC/258_v_260_edit_v_WT/P260-1_S9_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "#treat2.pe.sizes <- getPESizes(\"~/BE_paper_2022/data/ATAC/258_v_260_edit_v_WT/P260-2_S10_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "#treat3.pe.sizes <- getPESizes(\"~/BE_paper_2022/data/ATAC/258_v_260_edit_v_WT/P260-3_S11_merge_R1.fastq.trim.merged.nodup.no_chrM_MT.bam\")\n",
    "#gc()\n",
    "\n",
    "# plot\n",
    "#hist(treat1.pe.sizes$sizes) # repeat for all replicates and conditions\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3281a1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################\n",
    "# count BAM reads in, e.g. 300 bp windows\n",
    "counts <- windowCounts(pe.bams, width=300, param=param) # set width as desired from the fragment length distribution analyses\n",
    "\n",
    "# filter uninteresting features (windows) by local enrichment\n",
    "# local background estimator: 2kb neighborhood\n",
    "neighbor <- suppressWarnings(resize(rowRanges(counts), width=2000, fix=\"center\")) # change width parameter as desired\n",
    "wider <- regionCounts(pe.bams, regions=neighbor, param=param) # count reads in neighborhoods\n",
    "# filter.stat <- filterWindows(counts, wider, type=\"local\") # the filterWindows() function is deprecated and has been replaced by filterWindowsLocal(). This is an archived step.\n",
    "filter.stat <- filterWindowsLocal(counts, wider)\n",
    "counts.local.filt <- counts[filter.stat$filter > log2(2),] # threshold of 2-fold increase in enrichment over 2kb neighborhood abundance; change as desired\n",
    "\n",
    "###############################\n",
    "# count BAM background bins (for TMM normalization)\n",
    "binned <- windowCounts(pe.bams, bin=TRUE, width=10000, param=param)\n",
    "\n",
    "peak.counts.tmm <- peak.counts.filt\n",
    "peak.counts.tmm <- normFactors(binned, se.out=peak.counts.tmm)\n",
    "working.windows <- peak.counts.tmm "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "356ca765",
   "metadata": {},
   "outputs": [],
   "source": [
    "y <- asDGEList(working.windows)\n",
    "colnames(y$counts) <- c(\"control1\", \"control2\", \"control3\",\"treat1\", \"treat2\",\"treat3\")\n",
    "rownames(y$samples) <- c(\"control1\", \"control2\", \"control3\",\"treat1\", \"treat2\",\"treat3\")\n",
    "y$samples$group <- c(\"control\", \"control\",\"control\", \"treat\", \"treat\",\"treat\")\n",
    "design <- model.matrix(~0+group, data=y$samples)\n",
    "colnames(design) <- c(\"control\", \"treat\") # CONFIRM THAT THESE COLUMNS CORRECTLY ALIGN!!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2b75d25c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GRanges object with 75321 ranges and 9 metadata columns:\n",
       "        seqnames              ranges strand |     logFC     logCPM         F\n",
       "           <Rle>           <IRanges>  <Rle> | <numeric>  <numeric> <numeric>\n",
       "  51947     chr8   49168208-49168325      * |   3.15814 -0.7664801   28.7774\n",
       "  75268    chr13   64047696-64047716      * |  -5.64667 -1.2357133   29.3465\n",
       "  66792    chr11   81708005-81708123      * |   1.88425 -0.0476377   21.6997\n",
       "  36165     chr5 144332480-144332529      * |   3.89552 -1.1906214   24.4731\n",
       "  45698     chr7   33536270-33536343      * |  -3.15173 -1.0195345   21.3078\n",
       "    ...      ...                 ...    ... .       ...        ...       ...\n",
       "  96768     chrY   22981832-22981858      * |         0   -2.60503         0\n",
       "  96769     chrY   24605630-24605704      * |         0   -2.60503         0\n",
       "  96770     chrY   24610105-24610281      * |         0   -2.60503         0\n",
       "  96771     chrY   25059810-25060071      * |         0   -2.60503         0\n",
       "  96772     chrY   25879121-25879242      * |         0   -2.60503         0\n",
       "        num.tests num.up.logFC num.down.logFC      PValue       FDR   direction\n",
       "        <integer>    <integer>      <integer>   <numeric> <numeric> <character>\n",
       "  51947         1            1              0 6.05914e-07 0.0228190          up\n",
       "  75268         1            0              1 5.32031e-07 0.0228190        down\n",
       "  66792         1            1              0 3.18945e-06 0.0800774          up\n",
       "  36165         1            1              0 4.69778e-06 0.0884603          up\n",
       "  45698         1            0              1 1.93871e-05 0.1891287        down\n",
       "    ...       ...          ...            ...         ...       ...         ...\n",
       "  96768         1            0              0           1         1        none\n",
       "  96769         1            0              0           1         1        none\n",
       "  96770         1            0              0           1         1        none\n",
       "  96771         1            0              0           1         1        none\n",
       "  96772         1            0              0           1         1        none\n",
       "  -------\n",
       "  seqinfo: 24 sequences from an unspecified genome; no seqlengths"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "# stabilize dispersion estimates with empirical bayes\n",
    "y <- estimateDisp(y, design)\n",
    "fit <- glmQLFit(y, design, robust=TRUE)\n",
    "\n",
    "# testing for differentially-accessible windows\n",
    "results <- glmQLFTest(fit, contrast=makeContrasts(treat-control, levels=design))\n",
    "# head(results$table)\n",
    "rowData(working.windows) <- cbind(rowData(working.windows), results$table) # combine GRanges rowdata with differential statistics\n",
    "# working.windows@rowRanges\n",
    " \n",
    "# merge nearby windows\n",
    "# up to \"tol\" distance apart: 500 bp in this case; max merged window width: 5000 bp\n",
    "merged.peaks <- mergeWindows(rowRanges(working.windows), tol=500L, max.width=5000L)\n",
    "# summary(width(merged.peaks$region))\n",
    "# should merge some peaks; change as desired\n",
    "\n",
    "# use most significant window as statistical representation for p-value and FDR for merged windows\n",
    "tab.best <- getBestTest(merged.peaks$id, results$table)\n",
    " \n",
    "# concatenating all relevant statistical data for final merged windows (no redundant columns)\n",
    "final.merged.peaks <- GRanges(cbind(as.data.frame(merged.peaks$region), results$table[tab.best$rep.test, -4], tab.best[,-c(7:8)]))\n",
    " \n",
    "# sort by FDR\n",
    "final.merged.peaks <- final.merged.peaks[order(final.merged.peaks@elementMetadata$FDR), ]\n",
    "final.merged.peaks\n",
    "\n",
    "write.table(final.merged.peaks, \n",
    "            \"edit_v_WT_stim_windows_all.txt\", \n",
    "            sep=\"\\t\", quote=F, col.names=T, row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c73ad700",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_counts <- cbind(as.data.frame(merged.peaks$region),\n",
    "                            results$fitted.values[tab.best$rep.test,])\n",
    "\n",
    "peak_ROI <- filter(out_counts, ((seqnames == 'chr12') & (start ==9764776)))\n",
    "\n",
    "cre4_counts <- peak_ROI %>% select(start, control1, control2,control3,treat1,treat2,treat3) %>% tidyr::pivot_longer(!start, names_to = \"group\", values_to = \"count\")\n",
    "\n",
    "lib_info <- results$samples\n",
    "lib_info$sample <- rownames(lib_info)\n",
    "cre4_counts <- merge(data.table(cre4_counts),\n",
    "                     lib_info,\n",
    "                     by.x = c(\"group\"),\n",
    "                     by.y = c(\"sample\"))\n",
    "\n",
    "cre4_counts <- cre4_counts %>% mutate(group = replace(group, grepl('control', group), \"control\"),\n",
    "                                      group = replace(group, grepl('treat', group), \"treat\"))\n",
    "\n",
    "\n",
    "cre4_counts$counts_norm <- cre4_counts$count / (cre4_counts$lib.size / 1000000.0)\n",
    "grouped_counts <- cre4_counts %>% group_by(group) %>% summarise(mean = mean(counts_norm),\n",
    "                                                                sd = sd(counts_norm))\n",
    "\n",
    "\n",
    "write.table(grouped_counts,\n",
    "            'RE4_peak.out',sep='\\t',quote=FALSE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0ce47b2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "### compute the FDR for peaks in this region\n",
    "system(\"tail -n+2 edit_v_WT_stim_windows_all.txt | awk '{OFS=\\\"\\t\\\"}{print $1,$2,$3,$6,$12}' > edit_v_WT_stim_windows_all.bed\")\n",
    "system('bedtools intersect -a edit_v_WT_stim_windows_all.bed -b ROI.bed > ROI_windows.bed')\n",
    "windows <- fread('ROI_windows.bed')\n",
    "windows$FDR <- p.adjust(windows$V5, method = \"BH\", n = 21)\n",
    "write.table(windows,'ROI_windows.adjusted.bed',sep='\\t',quote=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "870af6c9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 21 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>FDR</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr12</td><td> 9764776</td><td> 9765449</td><td>-0.64668748</td><td>0.001648112</td><td>0.03461034</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9751938</td><td> 9752109</td><td>-1.15165853</td><td>0.005775780</td><td>0.06064569</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9435991</td><td> 9436078</td><td> 1.22281002</td><td>0.031728652</td><td>0.18123094</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9708290</td><td> 9708440</td><td>-1.08209989</td><td>0.034594045</td><td>0.18123094</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9741861</td><td> 9742058</td><td>-0.74342827</td><td>0.043150224</td><td>0.18123094</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9951138</td><td> 9951553</td><td> 0.85227567</td><td>0.135138149</td><td>0.47298352</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9455710</td><td> 9455792</td><td> 0.74791141</td><td>0.178818753</td><td>0.51390732</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9798095</td><td> 9798254</td><td>-0.61107678</td><td>0.195774216</td><td>0.51390732</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9674157</td><td> 9674343</td><td> 0.49954120</td><td>0.220499024</td><td>0.51449772</td></tr>\n",
       "\t<tr><td>chr12</td><td>10190902</td><td>10191070</td><td> 0.38254338</td><td>0.284378352</td><td>0.59719454</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9760865</td><td> 9761074</td><td>-0.25740984</td><td>0.370070411</td><td>0.70649806</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9702365</td><td> 9702501</td><td> 0.24122425</td><td>0.522957027</td><td>0.91517480</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9283586</td><td> 9284494</td><td> 0.32815179</td><td>1.000000000</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9447803</td><td> 9448389</td><td>-0.19356080</td><td>0.887511945</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9469515</td><td> 9469589</td><td>-0.14422223</td><td>0.801046520</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9647968</td><td> 9648340</td><td> 0.01677811</td><td>0.945633045</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9650762</td><td> 9651038</td><td> 0.08643596</td><td>0.762979049</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9758774</td><td> 9759073</td><td> 0.36143744</td><td>0.819012335</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9814144</td><td> 9814644</td><td>-0.09236852</td><td>1.000000000</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td> 9906867</td><td> 9907189</td><td>-0.12982761</td><td>0.649630750</td><td>1.00000000</td></tr>\n",
       "\t<tr><td>chr12</td><td>10212331</td><td>10213645</td><td>-0.26705931</td><td>1.000000000</td><td>1.00000000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 21 × 6\n",
       "\\begin{tabular}{llllll}\n",
       " V1 & V2 & V3 & V4 & V5 & FDR\\\\\n",
       " <chr> & <int> & <int> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t chr12 &  9764776 &  9765449 & -0.64668748 & 0.001648112 & 0.03461034\\\\\n",
       "\t chr12 &  9751938 &  9752109 & -1.15165853 & 0.005775780 & 0.06064569\\\\\n",
       "\t chr12 &  9435991 &  9436078 &  1.22281002 & 0.031728652 & 0.18123094\\\\\n",
       "\t chr12 &  9708290 &  9708440 & -1.08209989 & 0.034594045 & 0.18123094\\\\\n",
       "\t chr12 &  9741861 &  9742058 & -0.74342827 & 0.043150224 & 0.18123094\\\\\n",
       "\t chr12 &  9951138 &  9951553 &  0.85227567 & 0.135138149 & 0.47298352\\\\\n",
       "\t chr12 &  9455710 &  9455792 &  0.74791141 & 0.178818753 & 0.51390732\\\\\n",
       "\t chr12 &  9798095 &  9798254 & -0.61107678 & 0.195774216 & 0.51390732\\\\\n",
       "\t chr12 &  9674157 &  9674343 &  0.49954120 & 0.220499024 & 0.51449772\\\\\n",
       "\t chr12 & 10190902 & 10191070 &  0.38254338 & 0.284378352 & 0.59719454\\\\\n",
       "\t chr12 &  9760865 &  9761074 & -0.25740984 & 0.370070411 & 0.70649806\\\\\n",
       "\t chr12 &  9702365 &  9702501 &  0.24122425 & 0.522957027 & 0.91517480\\\\\n",
       "\t chr12 &  9283586 &  9284494 &  0.32815179 & 1.000000000 & 1.00000000\\\\\n",
       "\t chr12 &  9447803 &  9448389 & -0.19356080 & 0.887511945 & 1.00000000\\\\\n",
       "\t chr12 &  9469515 &  9469589 & -0.14422223 & 0.801046520 & 1.00000000\\\\\n",
       "\t chr12 &  9647968 &  9648340 &  0.01677811 & 0.945633045 & 1.00000000\\\\\n",
       "\t chr12 &  9650762 &  9651038 &  0.08643596 & 0.762979049 & 1.00000000\\\\\n",
       "\t chr12 &  9758774 &  9759073 &  0.36143744 & 0.819012335 & 1.00000000\\\\\n",
       "\t chr12 &  9814144 &  9814644 & -0.09236852 & 1.000000000 & 1.00000000\\\\\n",
       "\t chr12 &  9906867 &  9907189 & -0.12982761 & 0.649630750 & 1.00000000\\\\\n",
       "\t chr12 & 10212331 & 10213645 & -0.26705931 & 1.000000000 & 1.00000000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 21 × 6\n",
       "\n",
       "| V1 &lt;chr&gt; | V2 &lt;int&gt; | V3 &lt;int&gt; | V4 &lt;dbl&gt; | V5 &lt;dbl&gt; | FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| chr12 |  9764776 |  9765449 | -0.64668748 | 0.001648112 | 0.03461034 |\n",
       "| chr12 |  9751938 |  9752109 | -1.15165853 | 0.005775780 | 0.06064569 |\n",
       "| chr12 |  9435991 |  9436078 |  1.22281002 | 0.031728652 | 0.18123094 |\n",
       "| chr12 |  9708290 |  9708440 | -1.08209989 | 0.034594045 | 0.18123094 |\n",
       "| chr12 |  9741861 |  9742058 | -0.74342827 | 0.043150224 | 0.18123094 |\n",
       "| chr12 |  9951138 |  9951553 |  0.85227567 | 0.135138149 | 0.47298352 |\n",
       "| chr12 |  9455710 |  9455792 |  0.74791141 | 0.178818753 | 0.51390732 |\n",
       "| chr12 |  9798095 |  9798254 | -0.61107678 | 0.195774216 | 0.51390732 |\n",
       "| chr12 |  9674157 |  9674343 |  0.49954120 | 0.220499024 | 0.51449772 |\n",
       "| chr12 | 10190902 | 10191070 |  0.38254338 | 0.284378352 | 0.59719454 |\n",
       "| chr12 |  9760865 |  9761074 | -0.25740984 | 0.370070411 | 0.70649806 |\n",
       "| chr12 |  9702365 |  9702501 |  0.24122425 | 0.522957027 | 0.91517480 |\n",
       "| chr12 |  9283586 |  9284494 |  0.32815179 | 1.000000000 | 1.00000000 |\n",
       "| chr12 |  9447803 |  9448389 | -0.19356080 | 0.887511945 | 1.00000000 |\n",
       "| chr12 |  9469515 |  9469589 | -0.14422223 | 0.801046520 | 1.00000000 |\n",
       "| chr12 |  9647968 |  9648340 |  0.01677811 | 0.945633045 | 1.00000000 |\n",
       "| chr12 |  9650762 |  9651038 |  0.08643596 | 0.762979049 | 1.00000000 |\n",
       "| chr12 |  9758774 |  9759073 |  0.36143744 | 0.819012335 | 1.00000000 |\n",
       "| chr12 |  9814144 |  9814644 | -0.09236852 | 1.000000000 | 1.00000000 |\n",
       "| chr12 |  9906867 |  9907189 | -0.12982761 | 0.649630750 | 1.00000000 |\n",
       "| chr12 | 10212331 | 10213645 | -0.26705931 | 1.000000000 | 1.00000000 |\n",
       "\n"
      ],
      "text/plain": [
       "   V1    V2       V3       V4          V5          FDR       \n",
       "1  chr12  9764776  9765449 -0.64668748 0.001648112 0.03461034\n",
       "2  chr12  9751938  9752109 -1.15165853 0.005775780 0.06064569\n",
       "3  chr12  9435991  9436078  1.22281002 0.031728652 0.18123094\n",
       "4  chr12  9708290  9708440 -1.08209989 0.034594045 0.18123094\n",
       "5  chr12  9741861  9742058 -0.74342827 0.043150224 0.18123094\n",
       "6  chr12  9951138  9951553  0.85227567 0.135138149 0.47298352\n",
       "7  chr12  9455710  9455792  0.74791141 0.178818753 0.51390732\n",
       "8  chr12  9798095  9798254 -0.61107678 0.195774216 0.51390732\n",
       "9  chr12  9674157  9674343  0.49954120 0.220499024 0.51449772\n",
       "10 chr12 10190902 10191070  0.38254338 0.284378352 0.59719454\n",
       "11 chr12  9760865  9761074 -0.25740984 0.370070411 0.70649806\n",
       "12 chr12  9702365  9702501  0.24122425 0.522957027 0.91517480\n",
       "13 chr12  9283586  9284494  0.32815179 1.000000000 1.00000000\n",
       "14 chr12  9447803  9448389 -0.19356080 0.887511945 1.00000000\n",
       "15 chr12  9469515  9469589 -0.14422223 0.801046520 1.00000000\n",
       "16 chr12  9647968  9648340  0.01677811 0.945633045 1.00000000\n",
       "17 chr12  9650762  9651038  0.08643596 0.762979049 1.00000000\n",
       "18 chr12  9758774  9759073  0.36143744 0.819012335 1.00000000\n",
       "19 chr12  9814144  9814644 -0.09236852 1.000000000 1.00000000\n",
       "20 chr12  9906867  9907189 -0.12982761 0.649630750 1.00000000\n",
       "21 chr12 10212331 10213645 -0.26705931 1.000000000 1.00000000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "windows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "321be5ab",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "r-cpu.4-0.m76",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-0:m76"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
