{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c1f7707",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "library(tidyverse)\n",
    "library(Gviz)\n",
    "library(rtracklayer)\n",
    "library(GenomicFeatures)\n",
    "library(data.table)\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c5c9a4d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "GTF_dir <- \"../reference_files/gencode.v40.annotation.gtf\"\n",
    "\n",
    "#hg38_txdb <- makeTxDbFromGFF(GTF_dir,\n",
    "#                                 format=\"gtf\",\n",
    "#                                 organism=\"Homo sapiens\",\n",
    "#                                 dbxrefTag = \"gene_name\")\n",
    " \n",
    " \n",
    "#saveDb(hg38_txdb, \"~/BE_paper_2022/reference_files/hg38.txdb\")\n",
    "\n",
    "hg38_txdb <- loadDb(\"../reference_files/hg38.txdb\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "05d00890",
   "metadata": {},
   "outputs": [],
   "source": [
    "gtrack <- GenomeAxisTrack()\n",
    "# specify the start and end of chromosome \n",
    "chr_no <- 12 \n",
    "chr_start <- 9745000  # start of region\n",
    "chr_end <- 9770000 # end of region\n",
    "\n",
    "\n",
    "gen <- 'hg38'\n",
    "\n",
    "## set gene region track\n",
    "gtTrack <- GeneRegionTrack(hg38_txdb,\n",
    "                           chromosome=chr_no, # chromosome number\n",
    "                           start=chr_start, # start of region\n",
    "                           end=chr_end, # end of region\n",
    "                           transcriptAnnotation=\"gene_id\", # symbol is the gene symbol\n",
    "                           fontsize.group=20# free to adjust font size\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9cea849b",
   "metadata": {},
   "outputs": [],
   "source": [
    "h3k27ac_260 = import.bw(\"../data_and_analyses/ChIP/bigwigs/H3K27ac_p260.pval.bigwig\",as=\"GRanges\")\n",
    "h3k27ac_258 = import.bw(\"../data_and_analyses/ChIP/bigwigs/H3K27ac_p258.pval.bigwig\",as=\"GRanges\")\n",
    "bhlhe40_260 = import.bw(\"../data_and_analyses/ChIP/bigwigs/BHLHE40_p260_pooled.pval.bigwig\", as=\"GRanges\")\n",
    "bhlhe40_258 = import.bw(\"../data_and_analyses/ChIP/bigwigs//BHLHE40-P258-1_R1.srt.nodup_x_ctl.pooled.pval.signal.bigwig\", as=\"GRanges\")\n",
    "gata3_p260 = import.bw(\"../data_and_analyses/ChIP/bigwigs/GATA3_p260.pval.bigwig\", as=\"GRanges\")\n",
    "gata3_p258 = import.bw(\"../data_and_analyses/ChIP/bigwigs/GATA3_p258.pval.bigwig\", as=\"GRanges\")\n",
    "\n",
    "ctrl_bw_jurkat = import.bw(\"../data_and_analyses/ATAC/258_v_260_edit_v_WT/bigwigs/Jurkat_P258_stim.merged.bw\",as=\"GRanges\")\n",
    "stim_bw_jurkat = import.bw(\"../data_and_analyses/ATAC/258_v_260_edit_v_WT/bigwigs/Jurkat_P260_stim.merged.bw\",as=\"GRanges\")\n",
    "\n",
    "options(ucscChromosomeNames=FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef21de57",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "jurkat_ctrl_track <- DataTrack(\n",
    "  range = ctrl_bw_jurkat,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,100)\n",
    ")\n",
    "\n",
    "jurkat_stim_track <- DataTrack(\n",
    "  range = stim_bw_jurkat,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",,\n",
    "  ylim = c(0,100)\n",
    ")\n",
    "\n",
    "h3k27ac_260_tr <- DataTrack(\n",
    "  range = h3k27ac_260,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,400)\n",
    ")\n",
    "\n",
    "h3k27ac_258_tr <- DataTrack(\n",
    "  range = h3k27ac_258,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,400)\n",
    ")\n",
    "\n",
    "bhlhe40_260_tr <- DataTrack(\n",
    "  range = bhlhe40_260,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,30)\n",
    ")\n",
    "\n",
    "bhlhe40_258_tr <- DataTrack(\n",
    "  range = bhlhe40_258,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,30)\n",
    ")\n",
    "gata3_p260_tr <- DataTrack(\n",
    "  range = gata3_p260,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,300)\n",
    ")\n",
    "\n",
    "gata3_p258_tr <- DataTrack(\n",
    "  range = gata3_p258,\n",
    "  chromosome=chr_no,\n",
    "  start=chr_start, # start of region\n",
    "  end=chr_end, # end of region\n",
    "  col.histogram = \"red\",\n",
    "  ylim = c(0,300)\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a42f12a",
   "metadata": {},
   "outputs": [],
   "source": [
    "pdf(\"chip_atac_bigwig.pdf\")\n",
    "## modify remainder in AI \n",
    "plotTracks(list(gtrack,\n",
    "                gtTrack,\n",
    "                jurkat_ctrl_track,\n",
    "                jurkat_stim_track,\n",
    "                h3k27ac_258_tr,\n",
    "                h3k27ac_260_tr,\n",
    "                gata3_p258_tr,\n",
    "                gata3_p260_tr,\n",
    "                bhlhe40_258_tr,\n",
    "                bhlhe40_260_tr), shape = \"arrow\", \n",
    "           transcriptAnnotation = \"symbol\", chromosome = \"chr12\",from = 9740000, to= 9780000,\n",
    "           type = c(\"h\",\"coverage\")\n",
    "           )\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3c694b8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "environment": {
   "name": "r-cpu.4-0.m76",
   "type": "gcloud",
   "uri": "gcr.io/deeplearning-platform-release/r-cpu.4-0:m76"
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
